File: src/backend/internal/MidiSortingBuffer.cpp
#include "MidiSortingBuffer.h"
#include <vector>
#include <algorithm>
#include <iostream>

MidiSortingBuffer::MidiSortingBuffer() {
    messages.reserve(buffer_size);
}

uint32_t MidiSortingBuffer::n_events() const {
    return messages.size();
}

MidiStorageElem MidiSortingBuffer::get_event(uint32_t idx) const {
    if (dirty) {
        throw std::runtime_error("Access in merging buffer which is unsorted");
    }
    return messages[idx];
}

void MidiSortingBuffer::PROC_sort() {
    if (dirty) {
        std::stable_sort(messages.begin(), messages.end(), [](const MidiStorageElem &a, const MidiStorageElem &b) {
            return a.time < b.time;
        });
        dirty = false;
    }
}

void MidiSortingBuffer::PROC_process(uint32_t nframes) {
    PROC_sort();
}

void MidiSortingBuffer::PROC_prepare(uint32_t nframes) {
    PROC_clear();
}

void MidiSortingBuffer::PROC_clear() {
    messages.clear();
    dirty = false;
}

void MidiSortingBuffer::write_event(MidiStorageElem event) {
    if (event.size > 4) {
        throw std::runtime_error(
            "Midi merging buffer: message dropped because size > 4");
    }
    if (messages.size() >= messages.capacity()) {
        std::cerr << "Warning: expanded MIDI buffer on processing thread\n";
        messages.reserve(messages.size() * 2);
    }
    messages.push_back(event);
    dirty = true;
}

File: src/rust/backend_rust/src/midi_sorting_buffer.rs
//! MIDI sorting buffer implementation in Rust.
//!
//! A buffer that stores unordered MIDI messages, sorts them efficiently
//! and outputs them in order. Used for merging MIDI from multiple sources.

use crate::midi_storage::MidiStorageElem;

/// A MIDI buffer that can store and sort MIDI messages.
pub struct MidiSortingBuffer {
    /// Storage for MIDI messages
    messages: Vec<MidiStorageElem>,
    /// Whether the buffer needs sorting
    dirty: bool,
}

impl MidiSortingBuffer {
    /// Create a new MidiSortingBuffer with default capacity
    pub fn new() -> Self {
        MidiSortingBuffer {
            messages: Vec::with_capacity(1024),
            dirty: false,
        }
    }

    /// Get the number of events in the buffer
    pub fn n_events(&self) -> u32 {
        self.messages.len() as u32
    }

    /// Get an event by index
    /// # Panics
    /// Panics if buffer is dirty (needs sorting) - call sort first
    pub fn get_event(&self, idx: u32) -> MidiStorageElem {
        if self.dirty {
            panic!("Access in merging buffer which is unsorted");
        }
        self.messages[idx as usize].clone()
    }

    /// Sort the messages by time
    pub fn sort(&mut self) {
        if self.dirty {
            self.messages.sort_by(|a, b| a.time.cmp(&b.time));
            self.dirty = false;
        }
    }

    /// Clear the buffer
    pub fn clear(&mut self) {
        self.messages.clear();
        self.dirty = false;
    }

    /// Write an event to the buffer
    /// Returns false if the message was dropped (size > 4)
    pub fn write_event(&mut self, event: MidiStorageElem) -> bool {
        if event.size > 4 {
            return false;
        }
        self.messages.push(event);
        self.dirty = true;
        true
    }
}

impl Default for MidiSortingBuffer {
    fn default() -> Self {
        Self::new()
    }
}
